### PR TITLE
Better CLI

### PR DESCRIPTION
**Pull request by okonet**
**What kind of change does this PR introduce?**
feature / refactoring

**Did you add tests for your changes?**

No, we need to introduce e2e tests for that

**If relevant, did you update the documentation?**

Yes

**Summary**

Refactors the CLI code so we use commands for `migrate` and `init`. 
Made use of yargs to simplify the source code. 
Consolidated all webpack options into one file.

**Does this PR introduce a breaking change?**
Yes, the commands should be now called with `webpack-cli migrate` instead of `webpack-cli --migrate`

**Other information**

### CODE REVIEW

**Comment by ev1stensberg** [comment]
Ready for review @okonet ?

**Comment by okonet** [comment]
Almost there.

**Comment by okonet** [comment]
So after spending a good amount of time on this, I reached a dead end it seems. I can't think of a good way of doing this change as a non-breaking change to the existing webpack CLI.

The problem is that current CLI syntax looks like this `webpack <entry> [<entry>] <output>`. 

Making `migrate` and `init` tasks act as commands means, the syntax to use them will be `webpack init` which might be ambiguous to the current webpack CLI.

Another issue I tried to solve is to not display all options by default when running `webpack --help`. This also seems impossible with a non-breaking change.

A breaking change means an introduction of an additional command for webpack default behavior. For example `compile` or `start`.

There is probably a solution for this that involves parsing of arguments but I'm not sure yet how reliable this can be implemented. Any feedback is appreciated!

@sokra I think you should be involved in this discussion since implications can be pretty big.

**Comment by TheLarkInn** [comment]
Any good ideas @sokra

**Comment by okonet** [comment]
We discussed on Slack but I'm a bit drawn by work now so I'll continue on this in a week or so.

File: bin/webpack.js [hunk 1]
```diff
@@ -19,360 +19,22 @@ if (localCLI && path.relative(localCLI, __filename) !== '') {
...
-}
+yargs
+	.usage(`webpack-cli ${require('../package.json').version}
+
+Usage: $0 <cmd> [args]
+
+Usage without config file: webpack <entry> [<entry>] <output>
+Usage with config file: webpack
+`)
+	.epilogue('for more information see https://webpack.github.io/docs/cli.html')
```
Comment (pksjce) [inline bin/webpack.js]: This should point to new docs?

**Comment by DanielaValero** [comment]
@okonet anything here for me to help with?

**Comment by okonet** [comment]
More free time would be great :( Seriously though, having tests would help dramatically before I refactor even further. Not sure how much of this is covered.

**Comment by ev1stensberg** [comment]
@okonet we should get that react test suite up and running first, yea?

**Comment by okonet** [comment]
What react test suit? You mean Jest snapshot tests? yes, that would be great to have. We need 100% coverage if we don't want break current implementation.

**Comment by ev1stensberg** [comment]
@okonet You mentioned on Slack that instead of Soren, you could use this https://github.com/vadimdemedes/ink for testing

**Comment by okonet** [comment]
@ev1stensberg it's not for testing but for the CLI itself and I'd like to look into it.

**Comment by ev1stensberg** [comment]
@okonet gotcha! We would need to come up with a way to test E2E and produce results, both for the new features but also the "old" CLI commands that webpack have.

**Comment by ev1stensberg** [comment]
May be better to redo this in another fresh PR after its been merged with webpack, closing @okonet , reopen if you want to continue to work on this branch!

**Pull request closed without merging**